urllib3/poolmanager.py

```diff
@@ -52,26 +80,43 @@ def _default_key_normalizer(key_class, request_context):
     :param key_class:
         The class to use when constructing the key. This should be a namedtuple
         with the ``scheme`` and ``host`` keys at a minimum.
-
+    :type  key_class: namedtuple
     :param request_context:
         A dictionary-like object that contain the context for a request.
         It should contain a key for each field in the :class:`HTTPPoolKey`
+    :type  request_context: dict
     """
-    context = {}
-    for key in key_class._fields:
-        context[key] = request_context.get(key)
+    # Since we mutate the dictionary, make a copy first
+    context = request_context.copy()
     context['scheme'] = context['scheme'].lower()
     context['host'] = context['host'].lower()
+
+    # These are both dictionaries and need to be transformed into frozensets
+    for key in ('headers', '_proxy_headers', '_socks_options'):
+        if key in context and context[key] is not None:
+            context[key] = frozenset(context[key].items())
+
+    # Map the kwargs to the names in the namedtuple - this is necessary since
+    # namedtuples can't have fields starting with '_'.
+    for key in list(context.keys()):
+        context['key_' + key] = context[key]
+        context.pop(key)
```

Comment:
Since we're using `pop`, is there a reason not to do `content['key_' + key] = context.pop(key)`?